stdlib/public/Concurrency/CMakeLists.txt

```diff
@@ -35,19 +35,29 @@ endif()
 set(SWIFT_RUNTIME_CONCURRENCY_C_FLAGS)
 set(SWIFT_RUNTIME_CONCURRENCY_SWIFT_FLAGS)
 
+check_cxx_compiler_flag(-fswift-async-fp=always CXX_COMPILER_HAS_SWIFT_ASYNC_FP_FLAG)
+
 if(NOT swift_concurrency_async_fp_mode)
   set(swift_concurrency_async_fp_mode "always")
 endif()
 
-# Don't emit extended frame info on platforms other than darwin, system
-# backtracer and system debugger are unlikely to support it.
+# Don't emit extended frame info on platforms other than darwin, if the compiler
+# has that flag, as the system backtracer and system debugger are unlikely to
+# support it.
 if(CMAKE_SYSTEM_NAME STREQUAL Darwin)
-  list(APPEND SWIFT_RUNTIME_CONCURRENCY_C_FLAGS
-    "-fswift-async-fp=${swift_concurrency_async_fp_mode}")
-  list(APPEND SWIFT_RUNTIME_CONCURRENCY_SWIFT_FLAGS
-    "-Xfrontend"
-    "-swift-async-frame-pointer=${swift_concurrency_async_fp_mode}")
-else()
+  if(CXX_COMPILER_HAS_SWIFT_ASYNC_FP_FLAG)
+    list(APPEND SWIFT_RUNTIME_CONCURRENCY_C_FLAGS
+      "-fswift-async-fp=${swift_concurrency_async_fp_mode}")
+    list(APPEND SWIFT_RUNTIME_CONCURRENCY_SWIFT_FLAGS
+      "-Xfrontend"
+      "-swift-async-frame-pointer=${swift_concurrency_async_fp_mode}")
```

Comment:
The idea is to only use this block if the C++ compiler used on macOS supports this flag. That's not working?

Reply:
No, this is not going to work. We normally use the "just-built" compiler (a compiler build from the just checked out llvm-project)  to build the standard library's c++ files. Unconditionally checking the platform compiler is not the right thing.

Does Android set `SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER` to use that version of clang? If so you could check if `SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER` is true -- and only in the case it is true do compiler check logic.

Reply:
> No, this is not going to work. We normally use the "just-built" compiler (a compiler build from the checked out llvm-project) to build the standard library's c++ files. Unconditionally checking the platform compiler is not the right thing.

Oh, right, forgot about the weird way the native build works, as I mostly cross-compile.

> Does Android set SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER to use that version of clang?

No, I use the third option, [setting `--native-clang-tools-path` when cross-compiling the standalone stdlib for Android](https://github.com/buttaface/swift-android-sdk/blob/dba89d2a3fb1d027db6c47a261ef00e308560dfe/.github/workflows/sdks.yml#L134).

I'll look into this more. Sorry about the wild goose chase, @LucianoPAlmeida.

Reply:
> I'll look into this more. Sorry about the wild goose chase, @LucianoPAlmeida.

No worries, thanks for working on this :) 

> We normally use the "just-built" compiler (a compiler build from the checked out llvm-project) to build the standard library's c++ files. Unconditionally checking the platform compiler is not the right thing.

Hey @aschwaighofer

If I understand correctly that is true for the ninja builds which are working fine, the original problem we are facing see discussion [here](https://forums.swift.org/t/clang-error-unknown-argument-fswift-async-fp-always/) is that when building with `--xcode` the CXX compiler seems to be `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++` the one that ships with Xcode version installed which don't have this flag available yet. This happened already sometimes before with other flags for example: https://bugs.swift.org/browse/SR-14714 due the differences of the "just-built" from source clang compiler checked out from apple/llvm-project and the Xcode pre-built clang. 
I guess even though they are related to the same flag, the android and Xcode problems are slightly different issues. Does this makes sense?

Reply:
I understand. What we need to happen is to check the compiler that is used to compile the concurrency c++ sources for the availability of the flag.

Which compiler does the check check_cxx_compiler_flag use to perform the check on a standard ninja build? The just build one or the platform toolchain one?

Reply:
I would hope whichever one was chosen to build the stdlib, but I'm not sure.

Do you know where the Xcode build chooses to build the stdlib with the platform compiler and not the just-built one? I don't see where that's set.

Reply:
There is this flag `SWIFT_PREBUILT_CLANG` that is set based on something passed from somewhere else, but I'm not sure if that is the one doing this ...